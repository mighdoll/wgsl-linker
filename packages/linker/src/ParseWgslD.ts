import {
  anyNot,
  anyThrough,
  eof,
  fn,
  kind,
  matchingLexer,
  opt,
  or,
  Parser,
  ParserContext,
  ParserInit,
  preParse,
  repeat,
  req,
  seq,
  setTraceName,
  simpleParser,
  SrcMap,
  tracing,
  withSep,
} from "mini-parse";
import {
  AbstractElem,
  AliasElem,
  CallElem,
  FnElem,
  FnNameElem,
  GlobalDirectiveElem,
  StructElem,
  StructMemberElem,
  TypeNameElem,
  TypeRefElem,
  VarElem,
} from "./AbstractElems.js";
import { mainTokens } from "./MatchWgslD.js";
import { directive } from "./ParseDirective.js";
import {
  comment,
  makeElem,
  unknown,
  word,
  wordNumArgs,
} from "./ParseSupport.js";

/** parser that recognizes key parts of WGSL and also directives like #import */

// prettier gets confused if we leave the quoted parens inline so make consts for them here
const lParen = "(";
const rParen = ")";

export interface ParseState {
  ifStack: boolean[]; // stack used while processiing nested #if #else #endif directives
  params: Record<string, any>; // user provided params to templates, code gen and #if directives
}

const optAttributes = repeat(seq(kind(mainTokens.attr), opt(wordNumArgs)));
const possibleTypeRef = Symbol("typeRef");

const globalDirectiveOrAssert = seq(
  or("diagnostic", "enable", "requires", "const_assert"),
  req(anyThrough(";"))
).map((r) => {
  const e = makeElem<GlobalDirectiveElem>("globalDirective", r);
  r.app.state.push(e);
});

/** parse an identifier into a TypeNameElem */
export const typeNameDecl = req(word.named("name")).map((r) => {
  return makeElem<TypeNameElem>("typeName", r, ["name"]);
});

/** parse an identifier into a TypeNameElem */
export const fnNameDecl = req(word.named("name"), "missing fn name").map(
  (r) => {
    return makeElem<FnNameElem>("fnName", r, ["name"]);
  }
);

/** find possible references to user types (structs) in this possibly nested template */
export const template: Parser<any> = seq(
  "<",
  or(
    word.named(possibleTypeRef), // only the first element of the template can be a type
    fn(() => template)
  ),
  repeat(
    or(
      fn(() => template),
      anyNot(">") // we don't care about the rest of the template
    )
  ),
  req(">")
);

/** find possible references to user structs in this type specifier and any templates */
export const typeSpecifier: Parser<TypeRefElem[]> = seq(
  word.named(possibleTypeRef),
  opt(template)
).map((r) =>
  r.named[possibleTypeRef].map((name) => {
    const e = makeElem<TypeRefElem>("typeRef", r);
    e.name = name;
    return e;
  })
);

export const structMember = seq(
  optAttributes,
  word.named("name"),
  ":",
  req(typeSpecifier.named("typeRefs"))
).map((r) => {
  return makeElem<StructMemberElem>("member", r, ["name", "typeRefs"]);
});

export const structDecl = seq(
  "struct",
  req(typeNameDecl).named("nameElem"),
  req("{"),
  withSep(",", structMember).named("members"),
  req("}")
).map((r) => {
  const e = makeElem<StructElem>("struct", r, ["members"]);
  const nameElem = r.named.nameElem[0];
  e.nameElem = nameElem;
  e.name = nameElem.name;
  r.app.state.push(e);
});

// keywords that can be followed by (), not to be confused with fn calls
const callishKeyword = simpleParser("keyword", (ctx: ParserContext) => {
  const keywords = ["if", "for", "while", "const_assert", "return"];
  const token = ctx.lexer.next();
  const text = token?.text;
  if (text && keywords.includes(text)) {
    return text;
  }
});

export const fnCall = seq(
  word
    .named("name")
    .map((r) => makeElem<CallElem>("call", r, ["name"]))
    .named("calls"), // we collect this in fnDecl, to attach to FnElem
  "("
);

// prettier-ignore
const fnParam = seq(
  optAttributes,
  word,
  opt(seq(":", req(typeSpecifier.named("typeRefs"))))
);

const fnParamList = seq(lParen, withSep(",", fnParam), rParen);

// prettier-ignore
const variableDecl = seq(
  or("const", "var", "let", "override"), 
  word, 
  ":", 
  req(typeSpecifier).named("typeRefs")
);

const block: Parser<any> = seq(
  "{",
  repeat(
    or(
      callishKeyword,
      fnCall,
      fn(() => block),
      variableDecl,
      anyNot("}")
    )
  ),
  req("}")
);

export const fnDecl = seq(
  optAttributes,
  "fn",
  req(fnNameDecl).named("nameElem"),
  // TODO some built in functions can have a template here, e.g. bitcast
  req(fnParamList),
  opt(seq("->", optAttributes, typeSpecifier.named("typeRefs"))),
  req(block)
).map((r) => {
  const e = makeElem<FnElem>("fn", r);
  const nameElem = r.named.nameElem[0];
  e.nameElem = nameElem;
  e.name = nameElem.name;
  e.calls = r.named.calls || [];
  e.typeRefs = r.named.typeRefs?.flat() || [];
  r.app.state.push(e);
});

export const globalVar = seq(
  optAttributes,
  or("const", "override", "var"),
  opt(template),
  word.named("name"),
  opt(seq(":", req(typeSpecifier.named("typeRefs")))),
  req(anyThrough(";"))
).map((r) => {
  const e = makeElem<VarElem>("var", r, ["name"]);
  e.typeRefs = r.named.typeRefs?.flat() || [];
  r.app.state.push(e);
});

export const globalAlias = seq(
  "alias",
  req(word.named("name")),
  req("="),
  req(typeSpecifier).named("typeRefs"),
  req(";")
).map((r) => {
  const e = makeElem<AliasElem>("alias", r, ["name", "typeRefs"]);
  r.app.state.push(e);
});

const globalDecl = or(fnDecl, globalVar, globalAlias, structDecl, ";");

const rootDecl = or(globalDirectiveOrAssert, globalDecl, directive, unknown);

const root = preParse(comment, seq(repeat(rootDecl), eof()));

export function parseWgslD(
  src: string,
  srcMap?: SrcMap,
  params: Record<string, any> = {},
  maxParseCount: number | undefined = undefined,
  grammar = root
): AbstractElem[] {
  const lexer = matchingLexer(src, mainTokens);
  const state: AbstractElem[] = [];
  const context: ParseState = { ifStack: [], params };
  const app = {
    context,
    state,
  };
  const init: ParserInit = {
    lexer,
    app,
    srcMap,
    maxParseCount,
  };

  grammar.parse(init);

  return app.state;
}

if (tracing) {
  const names: Record<string, Parser<unknown>> = {
    globalDirectiveOrAssert,
    template,
    typeSpecifier,
    structMember,
    structDecl,
    fnCall,
    fnParam,
    fnParamList,
    block,
    fnDecl,
    globalVar,
    globalAlias,
    globalDecl,
    rootDecl,
    root,
  };

  Object.entries(names).forEach(([name, parser]) => {
    setTraceName(parser, name);
  });
}
