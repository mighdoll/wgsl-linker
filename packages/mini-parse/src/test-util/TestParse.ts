import { matchingLexer, matchOneOf, OptParserResult, Parser, TokenMatcher, tokenMatcher, _withBaseLogger, NameRecord } from "mini-parse";
import { expect } from "vitest";
import { logCatch } from "./LogCatcher.js";

const symbolSet =
  "& && -> @ / ! [ ] { } : , = == != > >= < << <= % - -- " + 
  ". + ++ | || ( ) ; * ~ ^ // /* */ += -= *= /= %= &= |= ^= >>= <<= <<";
export const testTokens = tokenMatcher({
  directive: /#[a-zA-Z_]\w*/,
  word: /[a-zA-Z_]\w*/,
  attr: /@[a-zA-Z_]\w*/,
  symbol: matchOneOf(symbolSet),
  digits: /\d+/,
  ws: /\s+/,
});

export interface TestParseResult<T, S = any, N extends NameRecord = never> {
  parsed: OptParserResult<T, N>;
  position: number;
  appState: S[];
}

/** utility for testing parsers */
export function testParse<T, S = any, N extends NameRecord = never>(
  p: Parser<T, N>,
  src: string,
  tokenMatcher: TokenMatcher = testTokens
): TestParseResult<T, S, N> {
  const lexer = matchingLexer(src, tokenMatcher);
  const app = {
    state: [],
    context: undefined,
  };
  const parsed = p.parse({ lexer, app, maxParseCount: 1000 });
  return { parsed, position: lexer.position(), appState: app.state };
}

/** run a test function and expect that no error logs are produced */
export function expectNoLogErr<T>(fn: () => T): T {
  const { log, logged } = logCatch();
  const result = _withBaseLogger(log, fn);
  expect(logged()).eq("");
  return result;
}
