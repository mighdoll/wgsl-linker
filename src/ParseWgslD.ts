import {
  ExportElem,
  ImportElem,
  StructElem,
  CallElem,
  FnElem,
  AbstractElem,
  ImportingItem,
} from "./AbstractElems.js";
import { matchingLexer } from "./MatchingLexer.js";
import {
  directiveArgsTokens,
  lineCommentTokens,
  mainTokens,
} from "./MatchWgslD.js";
import {
  ExtendedResult,
  ParserContext,
  ParserStage,
  any,
  eof,
  fn,
  kind,
  not,
  opt,
  or,
  repeat,
  seq,
  tokens,
} from "./ParserCombinator.js";

/** parser that recognizes key parts of WGSL and also directives like #import */

const m = mainTokens;
const a = directiveArgsTokens;
const l = lineCommentTokens;

/** ( <a> <,b>* ) */
const directiveArgs: ParserStage<string[]> = seq(
  "(",
  kind(a.word).named("words"),
  repeat(seq(",", kind(a.word).named("words"))),
  ")"
)
  .map((r) => r.named.words as string[])
  .traceName("directiveArgs");

const eol = or("\n", eof());

const importingElem = seq(kind(a.word), directiveArgs)
  .map((r) => ({ importing: r.value[0], args: r.value[1] } as ImportingItem))
  .traceName("importingElem");

export const importing = seq(
  "importing",
  seq(importingElem.named("elem"), repeat(importingElem.named("elem")))
)
  .map((r) => r.named["elem"] as ImportingItem[])
  .traceName("importing");

/** #export <foo> <(a,b)> <importing bar(a) <zap(b)>* > EOL */
// prettier-ignore
const exportDirective = seq(
  "#export",
  tokens(
    directiveArgsTokens,
    seq(
      opt(kind(a.word).named("name")), 
      opt(directiveArgs.named("args")), 
      opt(importing.named("importing")), 
      eol
    )
  )
)
  .map((r) => {
    // flatten 'args' by putting it with the other extracted names
    const e = makeElem<ExportElem>("export", r, ["name", "importing", "args"]);
    r.app.push(e);
  })
  .traceName("export");

/** #import foo <(a,b)> <from bar> <as boo> EOL */
const importDirective = seq(
  "#import",
  tokens(
    directiveArgsTokens,
    seq(
      kind(a.word).named("name"),
      opt(directiveArgs.named("args")),
      opt(seq("as", kind(a.word).named("as"))),
      opt(seq("from", kind(a.word).named("from"))),
      eol
    )
  )
)
  .map((r) => {
    // flatten 'args' by putting it with the other extracted names
    const named: (keyof ImportElem)[] = ["name", "from", "as", "args"];
    const e = makeElem<ImportElem>("import", r, named, []);
    r.app.push(e);
  })
  .traceName("import");

const ifDirective: ParserStage<any> = seq(
  "#if",
  tokens(
    directiveArgsTokens,
    seq(opt("!").named("invert"), kind(m.word).named("name"), eol)
  ).toParser((r) => {
    const { params } = r.appState as ParseState;
    const ifArg = r.named["name"]?.[0] as string;
    const invert = r.named["invert"]?.[0] === "!";
    const arg = !!params[ifArg];
    const truthy = invert ? !arg : arg;
    return ifBody(r, truthy);
  })
).traceName("#if");

const elseDirective = seq("#else", tokens(directiveArgsTokens, eol))
  .toParser((r) => {
    const { ifStack } = r.appState as ParseState;
    const ifState = ifStack.pop();
    if (ifState === undefined) console.warn("unmatched #else", r.start);
    return ifBody(r, !ifState);
  })
  .traceName("#else");

function ifBody(
  r: ExtendedResult<any>,
  truthy: boolean
): ParserStage<any> | undefined {
  const { ifStack } = r.appState as ParseState;
  ifStack.push(truthy);
  if (!truthy) return skipUntilElseEndif;
}

const endifDirective = seq("#endif", tokens(directiveArgsTokens, eol))
  .map((r) => {
    const { ifStack } = r.appState as ParseState;
    const ifState = ifStack.pop();
    if (ifState === undefined) console.warn("unmatched #endif", r.start);
  })
  .traceName("#endif");

export const directive = or(
  exportDirective,
  importDirective,
  ifDirective,
  elseDirective,
  endifDirective
).traceName("directive or");
// .trace();

/** // <#import|#export|any> */
export const lineComment = seq(
  "//",
  tokens(lineCommentTokens, or(directive, kind(l.notDirective)))
).traceName("lineComment");

// prettier-ignore
const skipUntilElseEndif = repeat(
  seq(
    or(
      lineComment, 
      seq(
        not("#else"), 
        not("#endif"),
        any()
      ), 
    )
  )
).traceName("skipTo #else/#endif");

const structDecl = seq(
  "struct",
  kind(m.word),
  "{",
  repeat(or(lineComment, seq(not("}"), any()))),
  "}"
).map((r) => {
  const e = makeElem<StructElem>("struct", r, ["name"]);
  r.app.push(e);
});

export const fnCall = seq(
  kind(m.word)
    .named("call")
    .map((r) => makeElem<CallElem>("call", r, ["call"]))
    .named("calls"), // we collect this in fnDecl, to attach to FnElem
  "("
);

const block: ParserStage<any> = seq(
  "{",
  repeat(
    or(
      lineComment,
      fnCall,
      fn(() => block),
      seq(not("}"), any())
    )
  ),
  "}"
).traceName("block");

export const fnDecl = seq(
  "fn",
  kind(a.word).named("name"),
  "(",
  repeat(or(lineComment, seq(not("{"), any()))),
  block
)
  .traceName("fnDecl")
  .map((r) => {
    const fn = makeElem<FnElem>("fn", r, ["name"]);
    fn.children = r.named.calls || [];
    r.app.push(fn);
  });

const unknown = any().map((r) => console.warn("???", r.value, r.start));

const rootDecl = or(fnDecl, directive, structDecl, lineComment, unknown);

const root = seq(repeat(rootDecl), eof());

interface ParseState {
  ifStack: boolean[];
  params: Record<string, any>;
}

export function parseWgslD(
  src: string,
  params: Record<string, any> = {}
): AbstractElem[] {
  const lexer = matchingLexer(src, mainTokens);
  const app: AbstractElem[] = [];

  const appState: ParseState = { ifStack: [], params };
  const context: ParserContext = { lexer, app, appState: appState };

  root(context);

  return context.app;
}

/** creat an AbstractElem by pulling fields from named parse results */
function makeElem<U extends AbstractElem>(
  kind: U["kind"],
  er: ExtendedResult<any>,
  named: (keyof U)[],
  namedArrays: (keyof U)[] = []
): U {
  const { start, end } = er;
  const nv = mapIfDefined(named, er.named as NameRecord<U>, true);
  const av = mapIfDefined(namedArrays, er.named as NameRecord<U>);
  return { kind, start, end, ...nv, ...av } as U;
}

type NameRecord<A> = Record<keyof A, string[]>;

function mapIfDefined<A>(
  keys: (keyof A)[],
  array: Record<keyof A, string[]>,
  firstElem?: boolean
): Partial<Record<keyof A, string>> {
  const entries = keys.flatMap((k) => {
    const ak = array[k];
    const v = firstElem ? ak?.[0] : ak;

    if (v === undefined) return [];
    else return [[k, v]];
  });
  return Object.fromEntries(entries);
}
