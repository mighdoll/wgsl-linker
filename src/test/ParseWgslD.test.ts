import { expect, test } from "vitest";
import { directive, lineComment, parseMiniWgsl } from "../ParseWgslD.js";
import { testParse } from "./TestParse.js";

// import { enableTracing } from "../ParserTracing.js";
// enableTracing();

test("parse empty string", () => {
  const parsed = parseMiniWgsl("");
  expect(parsed).toMatchSnapshot();
});

test("directive parses #export", () => {
  const parsed = testParse(directive, "#export");
  expect(parsed.app[0].kind).equals("export");
});

test("parse #export", () => {
  const parsed = parseMiniWgsl("#export");
  expect(parsed[0].kind).equals("export");
});

test("parse #export foo", () => {
  const parsed = parseMiniWgsl("#export foo");
  expect(parsed).toMatchSnapshot();
});

test("parse #export foo(bar)", () => {
  const parsed = parseMiniWgsl("#export foo(bar)");
  expect(parsed).toMatchSnapshot();
});

test("parse #export foo(bar, baz, boo)", () => {
  const parsed = parseMiniWgsl("#export foo(bar, baz, boo)");
  expect(parsed).toMatchSnapshot();
});

test("parse #import foo", () => {
  const parsed = parseMiniWgsl("#import foo");
  expect(parsed).toMatchSnapshot();
});

test("parse #import foo(a,b) as baz from bar", () => {
  const parsed = parseMiniWgsl("#import foo as baz from bar");
  expect(parsed).toMatchSnapshot();
});

test("lineComment parse // foo bar", () => {
  const src = "// foo bar";
  const { position } = testParse(lineComment, src);
  expect(position).eq(src.length);
});

test("lineComment parse // foo bar \\n", () => {
  const comment = "// foo bar";
  const src = comment + "\n x";
  const { position } = testParse(lineComment, src);
  expect(position).eq(comment.length);
});

test("lineComment parse // #export foo", () => {
  const src = "// #export foo";
  const { position, app } = testParse(lineComment, src);
  expect(position).eq(src.length);
  expect(app).toMatchSnapshot();
});

test("parse fn foo() { }", () => {
  const src = "fn foo() { }";
  const parsed = parseMiniWgsl(src);
  expect(parsed).toMatchSnapshot();
});

test("parse fn with calls", () => {
  const src = "fn foo() { foo(); bar(); }";
  const parsed = parseMiniWgsl(src);
  expect(parsed).toMatchSnapshot();
});

test("parse struct", () => {
  const src = "struct Foo { a: f32; b: i32; }";
  const parsed = parseMiniWgsl(src);
  expect(parsed).toMatchSnapshot();
});

test("parse fn with line comment", () => {
  const src = `
    fn binaryOp() { // binOpImpl
    }`;
  const parsed = parseMiniWgsl(src);
  expect(parsed).toMatchSnapshot();
});

test("parse #export(foo) with trailing space", () => {
  const src = `
    // #export (Elem) 
    `;
  const parsed = parseMiniWgsl(src);
  expect(parsed).toMatchSnapshot();
});

test("parse #if !foo #else #endif", () => {
  const src = `
    // #if !foo
      fn f() { notfoo(); }
    // #else
      fn f() { foo(); }
    // #endif 
    `;
  const parsed = parseMiniWgsl(src, { foo: true });
  console.log(parsed);
});
