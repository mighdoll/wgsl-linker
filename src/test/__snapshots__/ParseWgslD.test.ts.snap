// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`importing parses importing bar(A) fog(B) 1`] = `
[
  {
    "args": [
      "A",
    ],
    "end": 21,
    "kind": "import",
    "name": "bar",
    "start": 15,
  },
  {
    "args": [
      "B",
    ],
    "end": 29,
    "kind": "import",
    "name": "fog",
    "start": 23,
  },
]
`;

exports[`lineComment parse // #export foo 1`] = `
[
  {
    "end": 14,
    "kind": "export",
    "name": "foo",
    "start": 3,
  },
]
`;

exports[`parse #export foo 1`] = `
[
  {
    "end": 11,
    "kind": "export",
    "name": "foo",
    "start": 0,
  },
]
`;

exports[`parse #export foo(bar) 1`] = `
[
  {
    "args": [
      "bar",
    ],
    "end": 16,
    "kind": "export",
    "name": "foo",
    "start": 0,
  },
]
`;

exports[`parse #export foo(bar, baz, boo) 1`] = `
[
  {
    "args": [
      "bar",
      "baz",
      "boo",
    ],
    "end": 26,
    "kind": "export",
    "name": "foo",
    "start": 0,
  },
]
`;

exports[`parse #export(A, B) importing bar(A) 1`] = `
{
  "args": [
    "A",
    "B",
  ],
  "end": 36,
  "importing": [
    {
      "args": [
        "A",
      ],
      "end": 35,
      "kind": "import",
      "name": "bar",
      "start": 29,
    },
  ],
  "kind": "export",
  "name": "bar",
  "start": 5,
}
`;

exports[`parse #export(foo) with trailing space 1`] = `
[
  {
    "args": [
      "Elem",
    ],
    "end": 24,
    "kind": "export",
    "start": 8,
  },
]
`;

exports[`parse #import foo 1`] = `
[
  {
    "end": 11,
    "kind": "import",
    "name": "foo",
    "start": 0,
  },
]
`;

exports[`parse #import foo(a,b) as baz from bar 1`] = `
[
  {
    "as": "baz",
    "end": 27,
    "from": "bar",
    "kind": "import",
    "name": "foo",
    "start": 0,
  },
]
`;

exports[`parse @attribute before fn 1`] = `
[
  {
    "children": [],
    "end": 31,
    "kind": "fn",
    "name": "main",
    "start": 5,
  },
]
`;

exports[`parse @compute @workgroup_size(a, b, 1) before fn 1`] = `
[
  {
    "children": [],
    "end": 61,
    "kind": "fn",
    "name": "main",
    "start": 5,
  },
]
`;

exports[`parse alias 1`] = `[]`;

exports[`parse and ignore const_assert 1`] = `
[
  {
    "children": [],
    "end": 42,
    "kind": "fn",
    "name": "main",
    "start": 30,
  },
]
`;

exports[`parse and ignore global diagnostic 1`] = `
[
  {
    "children": [],
    "end": 61,
    "kind": "fn",
    "name": "main",
    "start": 49,
  },
]
`;

exports[`parse empty string 1`] = `[]`;

exports[`parse fn foo() { } 1`] = `
[
  {
    "children": [],
    "end": 12,
    "kind": "fn",
    "name": "foo",
    "start": 0,
  },
]
`;

exports[`parse fn with calls 1`] = `
[
  {
    "children": [
      {
        "call": "foo",
        "end": 14,
        "kind": "call",
        "start": 11,
      },
      {
        "call": "bar",
        "end": 21,
        "kind": "call",
        "start": 18,
      },
    ],
    "end": 26,
    "kind": "fn",
    "name": "foo",
    "start": 0,
  },
]
`;

exports[`parse fn with line comment 1`] = `
[
  {
    "children": [],
    "end": 39,
    "kind": "fn",
    "name": "binaryOp",
    "start": 5,
  },
]
`;

exports[`parse root level ;; 1`] = `[]`;

exports[`parse struct 1`] = `
[
  {
    "end": 30,
    "kind": "struct",
    "start": 0,
  },
]
`;

exports[`parse top level override and const 1`] = `
[
  {
    "children": [],
    "end": 56,
    "kind": "fn",
    "name": "main",
    "start": 44,
  },
]
`;

exports[`parse top level var 1`] = `
[
  {
    "children": [],
    "end": 76,
    "kind": "fn",
    "name": "main",
    "start": 64,
  },
]
`;

exports[`skipBlockComment parses /* comment */ 1`] = `
{
  "named": {},
  "value": [
    "/*",
    [
      {
        "kind": "word",
        "text": "comment",
      },
    ],
    "*/",
  ],
}
`;

exports[`wordNumArgs parses (a, b, 1) 1`] = `
[
  "a",
  "b",
  "1",
]
`;

exports[`wordNumArgs parses (a, b, 1) with line comments everywhere 1`] = `
[
  "a",
  "b",
  "1",
]
`;
